Testing audit server issue

Occasionally we get the follow error when running tests:

```sh
  1) Auditing Integration Tests TCP Audit Server should send TCP audit messages and save (valid):

      Uncaught AssertionError: expected 1 to be 2
      + expected - actual

      -1
      +2

      at Assertion.fail (node_modules/should/lib/assertion.js:92:17)
      at Assertion.Object.defineProperty.value (node_modules/should/lib/assertion.js:164:19)
      at Query.<anonymous> (test/integration/auditingIntegrationTests.coffee:134:37)
      at node_modules/mongoose/node_modules/kareem/index.js:177:19
      at node_modules/mongoose/node_modules/kareem/index.js:109:16



Warning:  Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.
```

--- a/test/integration/auditAPITests.coffee
+++ b/test/integration/auditAPITests.coffee
@@ -222,7 +222,7 @@ describe "API Integration Tests", ->
                       newAudits[1].participantObjectIdentification.length.should.be.exactly 1
                       newAudits[1].participantObjectIdentification[0].participantObjectID.should.be.exactly "https://localhost:8080/audits/#{result._id}"
                     done()
-                , 10 * global.testTimeoutFactor
+                , 100 * global.testTimeoutFactor
 
       it "should NOT generate an 'audit log used' audit when using basic (default) representation", (done) ->
         audit = new Audit auditData
@@ -328,7 +328,7 @@ describe "API Integration Tests", ->
                       newAudits[1].participantObjectIdentification.length.should.be.exactly 1
                       newAudits[1].participantObjectIdentification[0].participantObjectID.should.be.exactly "https://localhost:8080/audits/#{result._id}"
                     done()
-                , 10 * global.testTimeoutFactor
+                , 100 * global.testTimeoutFactor
 
 
     describe "*getAuditsFilterOptions", ->

--- a/test/integration/auditingIntegrationTests.coffee
+++ b/test/integration/auditingIntegrationTests.coffee
@@ -11,9 +11,7 @@ testAuditMessage = require('../unit/auditingTest').testAuditMessage
 describe "Auditing Integration Tests", ->
 
 
-  beforeEach (done) -> 
-    Audit.remove ->
-      server.start auditUDPPort: 5050, auditTlsPort: 5051, auditTcpPort: 5052, done
+  beforeEach (done) -> Audit.remove {}, -> server.start auditUDPPort: 5050, auditTlsPort: 5051, auditTcpPort: 5052, -> done()
 
   afterEach (done) -> server.stop -> done()
 
@@ -53,8 +51,9 @@ describe "Auditing Integration Tests", ->
 
       client = tls.connect 5051, 'localhost', options, ->
         messagePrependlength = "#{testAuditMessage.length} #{testAuditMessage}"
-        client.write messagePrependlength
-        client.end()
+        client.write messagePrependlength, (err) ->
+          return done err if err
+          client.end()
 
       client.on 'end', -> Audit.find {}, (err, audits) ->
         return done err if err
@@ -86,8 +85,9 @@ describe "Auditing Integration Tests", ->
     it "should send TCP audit messages and save (valid)", (done) ->
       client = net.connect 5052, 'localhost', ->
         messagePrependlength = testAuditMessage.length + ' ' + testAuditMessage
-        client.write messagePrependlength
-        client.end()
+        client.write messagePrependlength, (err) ->
+          return done err if err
+          client.end()
 
       client.on 'end', -> Audit.find {}, (err, audits) ->
         return done err if err
@@ -100,8 +100,9 @@ describe "Auditing Integration Tests", ->
     it "should send TCP audit message and NOT save (Invalid)", (done) ->
       client = net.connect 5052, 'localhost', ->
         # testAuditMessage does not have message length with space prepended
-        client.write testAuditMessage
-        client.end()
+        client.write testAuditMessage, (err) ->
+          return done err if err
+          client.end()
 
       client.on 'end', -> Audit.find {}, (err, audits) ->
         return done err if err

--- a/test/integration/auditingIntegrationTests.coffee
+++ b/test/integration/auditingIntegrationTests.coffee
@@ -41,46 +41,51 @@ describe "Auditing Integration Tests", ->
         setTimeout checkAudits, 100 * global.testTimeoutFactor
 
   describe "TLS Audit Server", ->
-
+  
     it "should send TLS audit messages and save (valid)", (done) ->
-
+  
       options =
         cert: fs.readFileSync "test/resources/trust-tls/cert1.pem"
         key:  fs.readFileSync "test/resources/trust-tls/key1.pem"
         ca: [ fs.readFileSync "test/resources/server-tls/cert.pem" ]
-
+  
       client = tls.connect 5051, 'localhost', options, ->
         messagePrependlength = "#{testAuditMessage.length} #{testAuditMessage}"
         client.write messagePrependlength, (err) ->
           return done err if err
           client.end()
-
-      client.on 'end', -> Audit.find {}, (err, audits) ->
-        return done err if err
-
-        # message fields already validate heavily in unit test, just perform basic check
-        audits.length.should.be.exactly 2 # 1 extra due to automatic actor start audit
-        audits[1].rawMessage.should.be.exactly testAuditMessage
-        done()
-
+  
+      client.on 'end', ->
+        checkAudits = -> Audit.find {}, (err, audits) ->
+          return done err if err
+        
+          # message fields already validate heavily in unit test, just perform basic check
+          audits.length.should.be.exactly 2 # 1 extra due to automatic actor start audit
+          audits[1].rawMessage.should.be.exactly testAuditMessage
+          done()
+        setTimeout checkAudits, 100 * global.testTimeoutFactor
+  
     it "should send TLS audit messages and NOT save (Invalid)", (done) ->
-
+  
       options =
         cert: fs.readFileSync "test/resources/trust-tls/cert1.pem"
         key:  fs.readFileSync "test/resources/trust-tls/key1.pem"
         ca: [ fs.readFileSync "test/resources/server-tls/cert.pem" ]
-
+  
       client = tls.connect 5051, 'localhost', options, ->
-        client.write testAuditMessage
-        client.end()
-
-      client.on 'end', -> Audit.find {}, (err, audits) ->
-        return done err if err
-
-        # message fields already validate heavily in unit test, just perform basic check
-        audits.length.should.be.exactly 1 # 1 extra due to automatic actor start audit
-        done()
-
+        client.write testAuditMessage, (err) ->
+          return done err if err
+          client.end()
+  
+      client.on 'end', ->
+        checkAudits = -> Audit.find {}, (err, audits) ->
+          return done err if err
+  
+          # message fields already validate heavily in unit test, just perform basic check
+          audits.length.should.be.exactly 1 # 1 extra due to automatic actor start audit
+          done()
+        setTimeout checkAudits, 100 * global.testTimeoutFactor
+  
   describe "TCP Audit Server", ->
     it "should send TCP audit messages and save (valid)", (done) ->
       client = net.connect 5052, 'localhost', ->
@@ -88,25 +93,29 @@ describe "Auditing Integration Tests", ->
         client.write messagePrependlength, (err) ->
           return done err if err
           client.end()
-
-      client.on 'end', -> Audit.find {}, (err, audits) ->
-        return done err if err
-
-        # message fields already validate heavily in unit test, just perform basic check
-        audits.length.should.be.exactly 2  # 1 extra due to automatic actor start audit
-        audits[1].rawMessage.should.be.exactly testAuditMessage
-        done()
-
+  
+      client.on 'end', ->
+        checkAudits = -> Audit.find {}, (err, audits) ->
+          return done err if err
+  
+          # message fields already validate heavily in unit test, just perform basic check
+          audits.length.should.be.exactly 2  # 1 extra due to automatic actor start audit
+          audits[1].rawMessage.should.be.exactly testAuditMessage
+          done()
+        setTimeout checkAudits, 100 * global.testTimeoutFactor
+  
     it "should send TCP audit message and NOT save (Invalid)", (done) ->
       client = net.connect 5052, 'localhost', ->
         # testAuditMessage does not have message length with space prepended
         client.write testAuditMessage, (err) ->
           return done err if err
           client.end()
-
-      client.on 'end', -> Audit.find {}, (err, audits) ->
-        return done err if err
-
-        # message fields already validate heavily in unit test, just perform basic check
-        audits.length.should.be.exactly 1 # 1 extra due to automatic actor start audit
-        done()
+  
+      client.on 'end', ->
+        checkAudits = -> Audit.find {}, (err, audits) ->
+          return done err if err
+  
+          # message fields already validate heavily in unit test, just perform basic check
+          audits.length.should.be.exactly 1 # 1 extra due to automatic actor start audit
+          done()
+        setTimeout checkAudits, 100 * global.testTimeoutFactor

--- a/test/integration/auditingIntegrationTests.coffee
+++ b/test/integration/auditingIntegrationTests.coffee
@@ -11,7 +11,9 @@ testAuditMessage = require('../unit/auditingTest').testAuditMessage
 describe "Auditing Integration Tests", ->
 
 
-  beforeEach (done) -> Audit.remove {}, -> server.start auditUDPPort: 5050, auditTlsPort: 5051, auditTcpPort: 5052, -> done()
+  beforeEach (done) -> 
+    Audit.remove ->
+      server.start auditUDPPort: 5050, auditTlsPort: 5051, auditTcpPort: 5052, done
 
   afterEach (done) -> server.stop -> done()
 

--- a/test/integration/mediatorAPITests.coffee
+++ b/test/integration/mediatorAPITests.coffee
@@ -22,7 +22,7 @@ describe "API Integration Tests", ->
           type: 'http'
         }
       ]
-      defaultChannelConfig: [
+      defaultChannelConfig: [{
         name: "Save Encounter 1"
         urlPattern: "/encounters"
         type: 'http'
@@ -35,7 +35,7 @@ describe "API Integration Tests", ->
             type: 'http'
           }
         ]
-      ,
+      }, {
         name: "Save Encounter 2"
         urlPattern: "/encounters2"
         type: 'http'
@@ -48,7 +48,7 @@ describe "API Integration Tests", ->
             type: 'http'
           }
         ]
-      ]
+      }]
 
     mediator2 =
       urn: "urn:uuid:25ABAB99-23BF-4AAB-8832-7E07E4EA5902"
@@ -1145,7 +1145,8 @@ describe "API Integration Tests", ->
               done()
 
       it 'should add all channels in the defaultChannelConfig property', (done) ->
-        new Mediator(mediator1).save ->
+        new Mediator(mediator1).save (err) ->
+          return done err if err
           request("https://localhost:8080")
             .post("/mediators/urn:uuid:EEA84E13-1C92-467C-B0BD-7C480462D1ED/channels")
             .set("auth-username", testUtils.rootUser.email)
@@ -1155,15 +1156,13 @@ describe "API Integration Tests", ->
             .send([])
             .expect(201)
             .end (err, res) ->
-              if err
-                done err
-              else
-                Channel.find {}, (err, channels) ->
-                  done err if err
-                  channels.length.should.be.exactly 2
-                  channels[0].name.should.be.exactly 'Save Encounter 1'
-                  channels[1].name.should.be.exactly 'Save Encounter 2'
-                  done()
+              return done err if err
+              Channel.find {}, (err, channels) ->
+                return done err if err
+                channels.length.should.be.exactly 2
+                channels[0].name.should.be.exactly 'Save Encounter 1'
+                channels[1].name.should.be.exactly 'Save Encounter 2'
+                done()
 
       it 'should add selected channels in the defaultChannelConfig property if the body is set (save one)', (done) ->
         new Mediator(mediator1).save ->

--- a/test/integration/mediatorAPITests.coffee
+++ b/test/integration/mediatorAPITests.coffee
@@ -1160,10 +1160,9 @@ describe "API Integration Tests", ->
               Channel.find {}, (err, channels) ->
                 return done err if err
                 channels.length.should.be.exactly 2
-                channelX = true if channels[0].name is 'Save Encounter 1' or 'Save Encounter 2'
-                channelY = true if channels[1].name is 'Save Encounter 1' or 'Save Encounter 2'
-                channelY.should.be true
-                channelX.should.be true
+                channelNames = channels.map (channel) -> channel.name
+                channelNames.should.containEql 'Save Encounter 1'
+                channelNames.should.containEql 'Save Encounter 2'
                 done()
 
       it 'should add selected channels in the defaultChannelConfig property if the body is set (save one)', (done) ->
@@ -1203,10 +1202,9 @@ describe "API Integration Tests", ->
                 Channel.find {}, (err, channels) ->
                   done err if err
                   channels.length.should.be.exactly 2
-                  channelX = true if channels[0].name is 'Save Encounter 1' or 'Save Encounter 2'
-                  channelY = true if channels[1].name is 'Save Encounter 1' or 'Save Encounter 2'
-                  channelY.should.be true
-                  channelX.should.be true
+                  channelNames = channels.map (channel) -> channel.name
+                  channelNames.should.containEql 'Save Encounter 1'
+                  channelNames.should.containEql 'Save Encounter 2'
                   done()
 
       it 'should return a 400 when a channel from the request body isn\'t found', (done) ->

--- a/test/integration/mediatorAPITests.coffee
+++ b/test/integration/mediatorAPITests.coffee
@@ -1160,8 +1160,10 @@ describe "API Integration Tests", ->
               Channel.find {}, (err, channels) ->
                 return done err if err
                 channels.length.should.be.exactly 2
-                channels[0].name.should.be.exactly 'Save Encounter 1'
-                channels[1].name.should.be.exactly 'Save Encounter 2'
+                channelX = true if channels[0].name is 'Save Encounter 1' or 'Save Encounter 2'
+                channelY = true if channels[1].name is 'Save Encounter 1' or 'Save Encounter 2'
+                channelY.should.be true
+                channelX.should.be true
                 done()
 
       it 'should add selected channels in the defaultChannelConfig property if the body is set (save one)', (done) ->
@@ -1201,8 +1203,10 @@ describe "API Integration Tests", ->
                 Channel.find {}, (err, channels) ->
                   done err if err
                   channels.length.should.be.exactly 2
-                  channels[0].name.should.be.exactly 'Save Encounter 1'
-                  channels[1].name.should.be.exactly 'Save Encounter 2'
+                  channelX = true if channels[0].name is 'Save Encounter 1' or 'Save Encounter 2'
+                  channelY = true if channels[1].name is 'Save Encounter 1' or 'Save Encounter 2'
+                  channelY.should.be true
+                  channelX.should.be true
                   done()
 
       it 'should return a 400 when a channel from the request body isn\'t found', (done) ->

--- a/test/integration/mediatorAPITests.coffee
+++ b/test/integration/mediatorAPITests.coffee
@@ -22,7 +22,7 @@ describe "API Integration Tests", ->
           type: 'http'
         }
       ]
-      defaultChannelConfig: [{
+      defaultChannelConfig: [
         name: "Save Encounter 1"
         urlPattern: "/encounters"
         type: 'http'
@@ -35,7 +35,7 @@ describe "API Integration Tests", ->
             type: 'http'
           }
         ]
-      }, {
+      ,
         name: "Save Encounter 2"
         urlPattern: "/encounters2"
         type: 'http'
@@ -48,7 +48,7 @@ describe "API Integration Tests", ->
             type: 'http'
           }
         ]
-      }]
+      ]
 
     mediator2 =
       urn: "urn:uuid:25ABAB99-23BF-4AAB-8832-7E07E4EA5902"

